vllm/engine/llm_engine.py

```diff
@@ -480,9 +480,12 @@ def _process_model_outputs(
             seq_group = scheduled_seq_group.seq_group
             seq_group.update_num_computed_tokens(
                 scheduled_seq_group.token_chunk_size)
-            # If uncomputed tokens > 0, it means prefill is chunked.
-            # We don't need to process outputs in that case.
-            if seq_group.get_num_uncomputed_tokens() == 0:
+
+            # If all sequences in the sequence group are in DECODE, then we can
+            # process the output tokens. Otherwise, they are (chunked) prefill
+            # samples and should not be processed.
+            stages = [seq.data._stage for seq in seq_group.seqs_dict.values()]
+            if all(stage == SequenceStage.DECODE for stage in stages):
```

Comment:
Answer for future readers:
* Chunked prefill and speculative decoding from a systems perspective are compatible, however the current vLLM implementations need work to be enabled together. I'll add a validation check which raises if both are enabled.
* The DECODE stage currently only reports 1 new token. This is used by the scheduler to prevent a batch from becoming compute-bound using the token budget. When chunked prefill is enabled, we will need to adjust this to take into account the "new tokens" computed during speculative verification and modify this value. When chunked prefill is disabled, the new token budget is max_num_batched_tokens, and we are OK with the fact that the budget system doesn't take speculative decoding into account.
* I'll make an issue for integrating chunked prefill with spec decode soon!